core/src/main/scala-2.13+/cats/data/NonEmptyLazyList.scala

```diff
@@ -62,14 +62,46 @@ object NonEmptyLazyList extends NonEmptyLazyListInstances {
   def fromSeq[A](as: Seq[A]): Option[NonEmptyLazyList[A]] =
     if (as.nonEmpty) Option(create(LazyList.from(as))) else None
 
-  def fromLazyListPrepend[A](a: A, ca: LazyList[A]): NonEmptyLazyList[A] =
-    create(a +: ca)
+  def fromLazyListPrepend[A](a: => A, ll: => LazyList[A]): NonEmptyLazyList[A] =
+    create(a #:: ll)
 
-  def fromLazyListAppend[A](ca: LazyList[A], a: A): NonEmptyLazyList[A] =
-    create(ca :+ a)
+  def fromLazyListAppend[A](ll: => LazyList[A], a: => A): NonEmptyLazyList[A] =
+    create(ll #::: a #:: LazyList.empty)
 
   def apply[A](a: => A, as: A*): NonEmptyLazyList[A] =
-    create(LazyList.concat(LazyList(a), LazyList.from(as)))
+    create(a #:: LazyList.from(as))
+
+  // allows the creation of fully lazy `NonEmptyLazyList`s by prepending to this
+  def maybe[A](ll: => LazyList[A]): Maybe[A] = new Maybe(() => ll)
+
+  final class Maybe[A] private[NonEmptyLazyList] (private[this] var mkLL: () => LazyList[A]) {
+    // because instances of this class are created explicitly, they might be
+    // reused, and we don't want to re-evaluate `mkLL`
+    private[this] lazy val ll = {
+      val res = mkLL()
+      mkLL = null // allow GC
+      res
+    }
+
+    def #::[B >: A](elem: => B): NonEmptyLazyList[B] =
+      create(elem #:: ll)
+    def #:::[B >: A](prefix: => NonEmptyLazyList[B]): NonEmptyLazyList[B] =
+      create(prefix.toLazyList #::: ll)
+    def #:::[B >: A](prefix: => LazyList[B]): Maybe[B] =
+      new Maybe(() => prefix #::: ll)
+  }
+
+  final class Deferrer[A] private[NonEmptyLazyList] (private val nell: () => NonEmptyLazyList[A]) extends AnyVal {
+    def #::[B >: A](elem: => B): NonEmptyLazyList[B] =
+      create(elem #:: nell().toLazyList)
+    def #:::[B >: A](prefix: => NonEmptyLazyList[B]): NonEmptyLazyList[B] =
+      create(prefix.toLazyList #::: nell().toLazyList)
+    def #:::[B >: A](prefix: => LazyList[B]): NonEmptyLazyList[B] =
+      create(prefix #::: nell().toLazyList)
```

Comment:
I didn't raelise `NonEmptyLazyList` was an opaque type. that's a bit frustrating, and causes a number of issues with these changes. not insurmountable ones—we'd just need to pick new names for a few methods—but it's still something that needs dealing with

Reply:
there's all sorts of things we could do: add an extra `:`, add an extra `#`, use `%` instead of `#` for one of them, etc.

Reply:
What about using a `DummyImplicit` to disambiguate?

Reply:
... that is a very obvious and simple solution. idk why I didn't think of it. cheers!